### PR TITLE
Rebase HIP 6.x + Update hip-tests

### PR DESCRIPTION
**Pull request by pvelesko**
Update HIP version from 5.1 to 6.x
* 45 new HIP APIs
* Placeholder code for cooperative groups - need to implement the following:

```
uint32_t __device__ __ockl_multi_grid_num_grids() {return 0;}; 
uint32_t __device__ __ockl_multi_grid_grid_rank() {return 0;}; 
uint32_t __device__ __ockl_multi_grid_size() {return 0;}; 
uint32_t __device__ __ockl_multi_grid_thread_rank() {return 0;}; 
uint32_t __device__ __ockl_multi_grid_is_valid() {return 0;}; 
uint32_t __device__ __ockl_multi_grid_sync() {return 0;}; 
uint32_t __device__ __ockl_grid_sync() {return 0;}; 
uint32_t __device__ __ockl_grid_is_valid() {return 0;}; 
void __device__ __builtin_amdgcn_fence(int, const char*){};
unsigned int __device__ __builtin_amdgcn_mbcnt_lo(unsigned int, unsigned int){return 0;};
unsigned int __device__ __builtin_amdgcn_read_exec(void){return 0;};
```

Update hip-tests submodule
* Switch to using hip-tests submodule by default
* Enable tests that we previously had commented out due to compilation issues - multiproc, stress, TypeQualifiers
* New test categories from the update -  perftests, performance

Update HIPCC
* Add support for `--genco` option. Reduces changes needed for upstreaming

The following individual tests are still excluded, mostly due to failures in LLVM-SPIRV-Translator
```
  coalesced_group.cc
  coalesced_groups_shfl_down_old.cc
  coalesced_groups_shfl_up_old.cc
  coalesced_group_tiled_partition.cc
```

The following test categores are excluded due to unsuported features.

Occupancy
```
# failure in template argument resolution
```

Surface - not implemented:
```
surf1DLayeredread
surf1Dread
surf2DLayeredread
surf2Dread
surf3Dread
surfCubemapread
```

Texture - not implemented:
```
tex1DGrad
tex1DLayered
tex1DLayeredGrad
tex1DLayeredLod
tex1DLod
tex3D
```

~~One final caveat: test discovery happens at runtime, every single time.. This means that to run a single test, we need to discover all the tests first. *Adds 70 seconds overhear for launching a single test*.~~
Total number of tests 975 -> 1898

### CODE REVIEW

**Comment by linehill** [comment]
> ```
> 863: CHIP warning [TID 3650729] [1710422994.385196816] : A device function is already registered and mapped to a different module.
> ```

Probably just a false positive. I wrote the warning at a time I didn’t fully understood what was happening (__hipRegisterFuntcion() called multiple times with same host pointer). I made a patch to remove it: https://github.com/CHIP-SPV/chipStar/pull/809.

**Comment by linehill** [comment]
FYI, building this branch from scratch with multiple jobs may end up in a situation where hipcc is not built before building hip-tests:
```
...
cd /mnt/md1/linehill/ws-chipstar-2/builds/chipstar/catch/catch_tests/unit/module && ../../../../bin/hipcc --genco --std=c++17 /mnt/md1/linehill/ws-chipstar-2/chipstar/hip-tests/catch/unit/module/get_function_module.cc -o get_function_module.code -I/mnt/md1/linehill/ws-chipstar-2/builds/chipstar/include/ --rocm-path=/opt/rocm
/bin/sh: 1: ../../../../bin/hipcc: not found
catch/catch_tests/unit/module/CMakeFiles/get_function_module.dir/build.make:72: recipe for target 'catch/catch_tests/unit/module/get_function_module.code' failed
make[2]: *** [catch/catch_tests/unit/module/get_function_module.code] Error 127
...
```

**Comment by linehill** [comment]
```
/mnt/md1/linehill/ws-chipstar-2/chipstar/include/hip/devicelib/type_casting_intrinsics.hh:32:36: warning: inline function '__double2hiint' is not defined [-Wundefined-inline]
extern "C++" inline __device__ int __double2hiint(double x);
```
Removing the `inline` suppresses the warning. The `inline` qualifier is only meaningful for function definitions.

**Review by linehill — CHANGES_REQUESTED** [review]
* Please run clang-format for the whole patch set. E.g. with `git clang-format main`.
* Did you check if there are new `hipDeviceProp_t` and `hipDeviceAttribute_t` entries and updated `hipGetDeviceProperties()` and `hipDeviceGetAttribute()` to respond to them?

File: include/hip/spirv_hip_runtime.h [hunk 1]
```diff
@@ -48,6 +48,30 @@ THE SOFTWARE.
...
+struct ihipEvent_t {};
+struct ihipCtx_t {};
+struct ihipStream_t {};
+struct ihipModule_t {};
+struct ihipModuleSymbol_t {};
+struct ihipGraph {};
+struct hipGraphNode {};
+struct hipGraphExec {};
+
+#define __managed__ __device__
```
[the remark this answers is not in the document]
Comment (pvelesko) [reply include/hip/spirv_hip_runtime.h:60]: without this we get undefined keyword so just added this as a placeholder to get everything to compile

[the remark this answers is not in the document]
Comment (linehill) [reply include/hip/spirv_hip_runtime.h:60]: Fair enough.

File: include/hip/spirv_hip_cooperative_groups_helper.h [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+uint32_t __device__ __ockl_multi_grid_grid_rank() {return 0;}; 
+uint32_t __device__ __ockl_multi_grid_size() {return 0;}; 
+uint32_t __device__ __ockl_multi_grid_thread_rank() {return 0;}; 
+uint32_t __device__ __ockl_multi_grid_is_valid() {return 0;}; 
+uint32_t __device__ __ockl_multi_grid_sync() {return 0;}; 
+uint32_t __device__ __ockl_grid_sync() {return 0;}; 
+uint32_t __device__ __ockl_grid_is_valid() {return 0;}; 
+void __device__ __builtin_amdgcn_fence(int, const char*){};
+unsigned int __device__ __builtin_amdgcn_mbcnt_lo(unsigned int, unsigned int){return 0;};
+unsigned int __device__ __builtin_amdgcn_read_exec(void){return 0;};
```
[the remark this answers is not in the document]
Comment (pvelesko) [reply include/hip/spirv_hip_cooperative_groups_helper.h:91]: None since we don't have an implementation. I added an entry

**Review by linehill — APPROVED** [review]
LGTM.

**Pull request merged**